src/dvc_task/proc/process.py

```diff
@@ -50,12 +51,17 @@ def asdict(self) -> Dict[str, Any]:
 
     def dump(self, filename: str) -> None:
         """Dump the process information into a file."""
-        with open(filename, "w", encoding="utf-8") as fobj:
-            lock(fobj, LOCK_EX)
-            try:
-                json.dump(self.asdict(), fobj)
-            finally:
-                unlock(fobj)
+        directory, file = os.path.split(filename)
+        with tempfile.NamedTemporaryFile(
+            mode="w",
+            encoding="utf-8",
+            dir=directory,
+            prefix=f"{file}.",
+            suffix=".tmp",
+            delete=False,
+        ) as tmp:
+            json.dump(self.asdict(), tmp)
+        os.replace(tmp.name, filename)
```

Comment:
How likely is it that this file gets overwritten? Is it only written once?

In Windows, `os.replace` may fail with PermissionError, if there are open file handles to destination, so for making it robust, we may need to retry.

Reply:
Very often, in `manager.py` we will update the process info via `__setitem__` if there are some status updates on queue tasks.

Reply:
@karajan1001, can you show me where we overwrite this? And when? And what do you mean by "very often", like how frequently that happens?

Apologies for asking too many questions, I'm still trying to understand the codebase. :)

Reply:
Although `os.replace()` can be fixed separately.

Reply:
OK, I'm sorry looking into it I found that I had mixed `ProcessInfo` with `ExecutorInfo`, here the `ProcessInfo` only dumps in https://github.com/iterative/dvc-task/blob/59caee678957668ad52cada7b78fc1f3ef3fa4b7/src/dvc_task/proc/process.py#L203